### ISSUE TITLE
[Heads up] Cloud currently down...

### ISSUE THREAD

**Issue by TheLogFather** [LLK/scratch-flash#1211]
No connection for https://scratch.mit.edu/projects/21501896/

Other projects are similarly not seeing the cloud values arrive after "connecting to server" box.

Also, cannot connect via API (so custom cloud clients are not working).

However, values are visible through varserver URL, and cloud logs look ok.

**Comment by TheLogFather** [comment]
UPDATE: cloud values for a project do arrive... eventually...
Taking a minute or two, though.

**Comment by TheLogFather** [comment]
Ah, fixed... great stuff! :)

**Comment by colbygk** [comment]
I have restarted the varserver for http.

**Comment by TheLogFather** [comment]
Cloud has been going through some rough patches today, on and off since ~1500UT.

It's currently really iffy – particularly new change getting reported back by the varserver URL.
See here for some real-time info:  https://scratch.mit.edu/projects/119629398/

**Comment by TheLogFather** [comment]
Looking good now...
(Was there a problem with a cloud server process? Or just high traffic?)

**Comment by TheLogFather** [comment]
Cloud is down again – for the last few hours (since ~1500UT)...

**Comment by TheLogFather** [comment]
OK, looks like it's come back again just now. :)

**Comment by TheLogFather** [comment]
Cloud down again for last few hours. Taking a minute or so again for projects to retrieve cloud values.
(Was also down for a few hours around midnight last night, GMT).

**Comment by TheLogFather** [comment]
Cloud was back ~1647UT, but was looking pretty dodgy from ~1730UT, basically unusable from ~1800UT, and failed soon after for a bit, until ~1840UT.

It was iffy again from ~1900UT, failing completely at ~1930UT, and is still down now...

**Comment by TheLogFather** [comment]
Cloud came back at 0223UT, and it's been basically OK since.

However, I'm seeing frequent latency spikes for the latest value getting reported back by the varserver URL (up to 1.5s delay about every 10-20 seconds).

I've seen this sort of behaviour in the past when cloud has been on the verge of failing, so it may be that it will be down again within the next hour or two...

**Comment by TheLogFather** [comment]
I think it's getting worse, so perhaps it'll be even less than hour...

**Comment by colbygk** [comment]
Looking at the servers (the software), they appear to be making normal progress and we may be running up against limits of the hardware in terms of where they are running. Further analysis is being tracked in another repo.

**Comment by colbygk** [comment]
FYI, last month we turned on a global throttling mechanism for certain classes of requests that arrive via http. By far, the type of query caught by the throttler are too many requests being sent too quickly to the cloud vars server. The response is a 429 to those requests.

**Comment by jwzimmer** [comment]
I'm not sure if the issues with the cloud reported here could be causing freshdesk no. 60047?

**Comment by colbygk** [comment]
Based on that one paragraph, it sounds to me like it's not so much that data is getting some how dropped for them, but that the editor itself is treating the variable differently or that the variable goes missing in the editor.
Perhaps there is some failure mode where if the cloud var server is not responding soon enough that the editor does something like dropping the var outright in the editor, but I don't think it would/should..  thoughts @cwillisf ?

**Comment by TheLogFather** [comment]
It looks like cloud is responding somewhat more sensibly now, after going down for about a minute at 1624UT.

**Comment by TheLogFather** [comment]
BTW, what I'm looking at is not in a project, but in a cloud client. Every two seconds it updates a cloudvar value, and then checks with the project's varserver URL to see if the new value is 'visible' there.

It checks the varserver URL about every 0.15s (up to 5s max) after setting the new value, until it finds the new value has appeared. (Is that too often?)

**Comment by colbygk** [comment]
Every 0.15s for 5 seconds is not too often

**Comment by TheLogFather** [comment]
OK, that's good to know – though it's very rare that it goes up to 5s at that rate.

Looking at the stats, if cloud is behaving well then the cloud client is typically doing three (occasionally a few more) cloudvar sets every 2 secs, one of which is followed 0.15s later by a single varserver URL check, sometimes followed by another 0.15s later if the first didn't show the update. (And every few minutes or so there's a 'lag spike' and it takes a few more checks, separated by 0.15s.)

I guess the downside of the way this works is that if cloud is already not behaving well for some reason then it'll end up sending more of those varserver requests, which is probably not ideal for cloud. :/

Given above, I've made my client now increase the delay between each new varserver request by 0.05s after the first two checks.

BTW: I'm currently getting long delays of one to four secs again for the varserver URL to notice updates.

**Comment by TheLogFather** [comment]
Oops, now it look like cloud has gone... :(

**Comment by TheLogFather** [comment]
Cloud is currently looking iffy, as far as varserver URLs are concerned – getting delays of one to four seconds before updates are seen through varserver URL.

However, the behaviour of the rest of cloud appears not too bad (projects are getting their values within a second or two after first loading, and the round-trip speed is at about a second, which is higher than the usual ~250ms for me, but still basically workable for most projects).

**Comment by jwzimmer** [comment]
Another user reported running into something that seems like it was caused by this issue in freshdesk no. 59768:
> I have been on Scratch many years on my normal account, and have used them many times. On this test account, I test my multiplayer games on my normal account and I make animations. Today I was going to make a voting thing, and I know how to make them. I did everything, then at the top it said: "Cloud variables only store numbers; see FAQ" bla bla bla... But the variable was not created. For the split second it has some loading thing, so it is an error after making it. Also, I tried on my other internet browser, Microsoft Edge, and that did not work either. I don't know if the servers have come down, or what.

**Comment by TheLogFather** [comment]
Looks like cloud is totally down at the moment – cloudvars not arriving at projects even several minutes after project loads. Also unable to create a viable cloud session via API.

**Comment by colbygk** [comment]
It does appear that the tcp version might be stuck, while the http server is processing ~475 successful requests per second at the moment. Restarting tcp.

**Comment by TheLogFather** [comment]
OK, that explains why the cloudvars were eventually arriving at the project a few mins after loading – once the port 531 connection timed out, and the project falls back to polling.

But it's interesting... for a while after you restarted, I was not seeing any delay spikes between a cloudvar getting set and the project's varserver URL reporting the updated value (i.e. within ~0.2s).

But I'm already now seeing delay spikes of 0.7 to even >1.5 secs happen two or three times per minute.

And on my cloud roundtrip project (https://scratch.mit.edu/projects/96491582/) I'm often seeing round-trip times >1 second.

**Comment by TheLogFather** [comment]
UPDATE to above: looks like it's all gone OK again now, since typing above...
(Round-trip times gone back down to 250-350ms, and not seeing varserver URL delays that often.)
Well... ok, then... :/

**Comment by TheLogFather** [comment]
Cloud is looking pretty messed up again already, unfortunately... :(

**Comment by TheLogFather** [comment]
Cloud came back at 0940UT, but now down again (since ~1225UT, though varserver URL response was looking highly dodgy from 1147UT, and heading that way from ~1120UT).

**Comment by TheLogFather** [comment]
Cloud is just about dead at the moment – getting varserver delay times of several seconds.

Can see real-time varserver delay on https://scratch.mit.edu/projects/119629398/
(See the "Cloud varserver 'set' delay" readout. Also, watching how the 'last contact time' lags behind true GMT/UTC gives an indication of the varserver delay.)

**Comment by TheLogFather** [comment]
Cloud is looking more stable today (just a brief 'wobble' around 0850-0900UT).  :)
Any changes to how it works? Or just less traffic...?

**Comment by TheLogFather** [comment]
Oops... not good now... :(
(Started looking a bit dodgy from ~1350UT, pretty much total meltdown ~1415UT.)

**Comment by algmwc5** [comment]
@TheLogFather It's back on
I don't know when it's back on

**Comment by TheLogFather** [comment]
It came back at 1647UT that day, but it's been up and down a several times since then. (E.g. ~1730UT was starting look iffy, then failing at ~1800UT until 1933UT. But only staying reasonable until ~2000UT, and failing ~2019UT until 2110UT, then iffy again by 2200UT, and failing ~2215UT until ~2300UT).

Basically, the behaviour is quite erratic, and when it's back up there's no guarantee it'll stay working well for more than an hour or two, if it's a busy time of day.  :(

Right now it's been mostly OK since ~0411UT (with a bad 'wobble' around 0820-0825UT).

**Comment by TheLogFather** [comment]
Cloud looking great for the past nearly 12hrs:  https://scratch.mit.edu/projects/119629398
Has anything changed, or is it just a big reduction in traffic/usage?

**Comment by TheLogFather** [comment]
Oh... oops...
Should've kept quiet.
Cloud has deteriorated steadily in the last hour or so.  :(
Curiously, it's practically the same timing as yesterday – working OK for hours until ~1400UT.

**Comment by thisandagain** [comment]
Hi @TheLogFather , our Cloud Variables infrastructure is simply unable to keep up with the load that we experience generally between noon and 6pm ET. While we are aware of the problem and have monitoring in-place there is little that we can do about it with the current architecture. We are currently in the process of redesigning, but it will take us some time before we can finalize design, implement, and deploy the new system. We hope to have updates for everyone soon. In the meantime I am going to close this issue.

**Comment by encloinc** [comment]
:/

**Comment by encloinc** [comment]
Could it be that the cloud server could be getting malicious spam requests from multiple API users?

**Comment by colbygk** [comment]
No. We have API limiters in place to avoid this. We are aware of the underlying issues behind these problems and are working towards a solution.

**Comment by TheLogFather** [comment]
@encloinc: I suspect the reason cloud struggles at certain times is likely because Scratchers will tend to naively write multiplayer projects which set various cloudvars at every pass through their loops. This probably means 10, 20 or maybe 30 cloud server requests every second from each person currently playing that project. And then the server has to send out all of those (non-self) changes back to all those players (and anyone else who happens to be on that project, even if not playing it).

Looking at your own cloud multiplayer engine, it seems as though you've done that yourself...

I wrote a couple of rants about it [here](https://scratch.mit.edu/discuss/post/2263220/) and [here](https://scratch.mit.edu/discuss/post/2263203/) (but hopefully with some helpful ideas for other ways to write multiplayer scripts).

**Comment by encloinc** [comment]
I highly doubt so @TheLogFather . A well done multiplayer project consists of a "Ping", this ping is constantly changed so that the other clients online, or incoming users, know who is online and their current place in the array (also known as an ID). This is inevitable because if this was not included, the ping would be governed by something more unreliable such as a x postion or a y position. We should not rely on the users inputs as a form of revealing if a client is online or not (as you suggested in your "rants") However, that said, something that could affect the cloud is perpetual scratch servers, for example a scratch client not limited by the restrictions of the API constantly changing a cloud variable.

**Comment by TheLogFather** [comment]
@encloinc: Yes, I'm aware of the 'ping', and the need to keep it changing, and I wouldn't suggest using something like X/Y pos to check player status (though it's the kind of thing I'd expect a relatively inexperienced Scratcher to do).

The point I'm trying to make is about changing the cloudvar(s) at every pass through a loop, i.e. something like 10-20x per second. This effectively 'spams' the cloud server.

The way I've written these things in the past (none that I've ever bothered to make public) is to keep to fewer than four or five changes per second, and find some way to 'interpolate' the movement to keep it looking relatively smooth.

Unfortunately, this kind of thing is clearly not the way an inexperienced Scratcher will think to write it, so cloud will tend to suffer for it at the moment... :(

**Comment by CosmicWebServices** [comment]
@TheLogFather you should make a project explaining this and comparing the right way to do this.

**Comment by encloinc** [comment]
@TheLogFather , interesting concept. The only flaw I can seem to find is that since the loop has been slowed down 4x, the ping will be sent at a much slower rate, decreasing the amount of times a client can change their ping (because of the 20x times a cloud var is changed, only about 10-15 are successful changes), this would mean increasing the amount of time it would take to kick an inactive player out, which is an issue. Another issue is that it isn't "real time multiplayer", it's an estimate on where the player will be in a ceartin period of time. However I do understand the reasoning behind this concept, I guess I'm going to have to ponder about it more.

**Comment by TheLogFather** [comment]
I tried to give some simple examples/ideas within that 'rant' (without getting into the whole 'ping' thing).

And, yes, you can't create a truly 'real-time multiplayer' game on Scratch... there's likely to be something like at least 0.2s latency between a player setting a cloudvar and the other players seeing that change – and there are occasionally latency spikes of even 1-2 seconds.

That means it's a good idea to design a multiplayer project to be able to function as only '*near*-real-time'. (It's worth looking at griffpatch's multiplayer stuff, and you'll see that he often even keeps a 'record' of the last few positions of the players, to be able to do the smoother movement – and that introduces a latency of even up to a second or so.)

Also, latency is only as low as 0.2s for players who are on the faster port 531 connection (and have decent 'net connections, which may not be the case in some locations). Anyone who could not make such a connection, due to firewall constraints, or who has randomly fallen back to polling (which can happen anytime from a few mins to a few hours of loading a project) will find themselves with at least a 1 second latency. :(

Kicking out a player is something that needs to be done really carefully – it's easy to get it wrong just due to random latency issues (remember the latency spikes). So such kicking off should only be done after several seconds of inactivity (which means even reducing 'ping' checking to only 2-3x per second would still be fine).

I do have a multiplayer base project that I made ages ago. Unfortunately, it's designed around an external 'game server' written in python, and I'm reluctant to hand that out at the moment for various reasons...

**Comment by encloinc** [comment]
Exactly, the engine I developed kicks the player out after 3 seconds, and it's governed by different clients (mostly id1 and 2). If we decrease the ping, ill have to up the wait time from something simple like 3 seconds to something massive like 12 seconds.

**Comment by TheLogFather** [comment]
I'd say 3 secs is a little on the low side. I've seen latency spikes of nearly that long even when cloud is working fairly well.

I have it set to 5 secs in mine, and I think that's reasonable – it's highly unlikely to be a latency spike at that length (if cloud is working well), and it's not that long.

I don't see why it would need to be anywhere near 12 secs...?

**Comment by encloinc** [comment]
Well, a way I see of solving this is that we create a multiplayer engine that's easy to use and make it used by a wide variety of people (that uses estimations), so people will ditch the old spam method, effectively fixing the cloud servers

**Comment by TheLogFather** [comment]
Yeah – except it really is actually rather hard to make a good, reliable multiplayer engine. And, especially, all the while, keeping it simple to understand and use.

I tried a couple of years ago, and got fed up with it...

Dealing with potential race conditions when kicking off players, and when players are joining, due to cloud latency and other oddities, is actually quite hard. If cloud is working nicely for everyone then these things are fairly straightforward and work nicely. But I found that it only takes one person with a 'bad' connection to mess it up for everyone else currently playing (or, at least, for *someone* else).


For example, there's an 'interesting situation' that can arise with a project's cloud connection under certain conditions where it ends up with a one-way, *write-only* connection. This has the unfortunate consequence that the project in this situation will happily try to 'kick off' everyone else (because, as far as it is concerned, everyone else has stopped pinging), while everyone else still sees this player's cloudvar changes, and so continues to accept them. (When existing players' cloudvars are getting zeroed like that by a rogue project, it can also mean a new player trying to join may think these slots are 'free'...)

There's also an opposite situation where a project can end up with a read-only connection – it sees everyone else's changes, but none of its own get through. And yet (and here's the important thing) *it has no way to know* that it's own changes are not getting through, because it sees all its own cloudvar changes in just the way it expects (since local changes to a cloudvar are made locally).

All of above are reasons why an external 'game server' is really much more reliable than trying to get a project to do it all – though it has the major downside that it's no longer Scratch...  :(

**Comment by encloinc** [comment]
Well what I did is that I made up a "rule book" to my engine. Basically, if the ID is greater than 2 and 2 and 1 are down at the same time, it will kick itself out, only id 1 can kick other players out, but id two removes the request after 4 seconds. The only way that this can fail is if I'd 2 enters write or read mode, because if ID 1 enters read mode or write mode, it'll be replaced by ID 2, but if ID 2 enters read or write mode, it'll try to kick ID 1 off, resulting in chaos. So now it's really just about detecting when ID 2 enters read or write mode, which we can use your "instant http fallback" detection method for it to detect if it's in fallback or if it's in read or write mode

**Comment by TheLogFather** [comment]
There's no way to detect read-only mode for sure unless another user sends back a response to your changes (which you can then test for). –And that's not possible if you're currently the only player, so you *really* cannot tell in that case.

And if you (think you) are the only player, and you're actually in write-only mode but don't know it, then when another player joins, you would never know they have joined, and you still cannot know which mode you are in (read-only, write-only, or read-write).

So that all means if the project thinks it is currently the only player in the game, it has no way to know if it is correct about that, or if it is sending changes which are ruining other players it cannot see (write only), or if it is totally invisible to other players who may be trying to join on its slot (read only) – unless, in that last case, it happens to 'catch' a value on its own cloudvar that it knows it did not set it to (and being able to catch that sort of thing actually requires that you *don't* set your own cloudvar too frequently!)

----

I haven't looked at exactly how your engine works, so can't say for sure, but it'd be worth you considering how situations such as those above will play out...

It'd also be worth you working through how your engine would deal with even the not-so-weird connection cases, where there are potential race conditions. For example, two users trying to join at very nearly the same time (on the same 'slot'?), but one or both have slightly bad/slow 'net connection for some reason (e.g. frequent latency of over half a second before seeing updates from others). –Is it possible both users could miss that the other is also trying to join on the same slot (so causing some chaos)?

----

Now, you brought up the fallback polling detection... I came across that capability more recently – after I got fed up trying to make a reliable (yet easy-to-use) multiplayer engine – and you rightly point out that it changes things (since I think the write-only case can only happen in polling mode).
It could be worth revisiting this with that in mind...

----

Anyway, all of above seems to be somewhat off-topic now, so I'll just say (at the risk of ruining it!) that I'm really pleased to see how good cloud has been looking over the last ~30hrs. (Has anything changed?)

**Comment by CosmicWebServices** [comment]
Yay!

**Comment by griffpatch** [comment]
I agree with @DadOfMrLog - I bet 99% of all scratchers trying out cloud
variables will actually be setting 3 or 4 cloud vars 30 times a second
(using one for x, one for y, one for direction, one for health) all being
set every cycle of a main loop. That can't be good for the cloud servers!
;) - I wonder how many of these little projects are actually running at
once over the entire scratch community? I don't think many cloud games are
actively played on scratch due to the unlikely event of actually finding
another player on a game when you are.

Sorry to prolong an off-topic thread! ;) but...

I see all the cloud network issues you talked about there :) my older cloud
engines assumed bad connections and therefore had a minimum broadcast
period of around 0.8 seconds. I would record all player movements in that
time and broadcast as a double buffered stream. This made game play very
smooth, but incurred a lag of more than 1.5 seconds a lot of the time! Was
rubbish for real time play, but most people didn't notice due to the types
of games it was designed for. Lately my cloud engines have been much more
akin to what @DadOfMrLog described where I only send data every 6-7 frames
(5 times a second), but I interpolate the other data to smooth things out.
I also detect network issues by counting the times that the player count
drops down to 1 and then goes back up. When there is only one player I drop
the broadcast frequency to once every second to compensate and allow
messages to propagate even under cloud fallback mode. This was the player
count will be allowed to go back up, but then it might fail again... so
after a few fails I ask for a project reload (which often is what is
required to get the players connected correctly again).

I've been wondering about scratch 3 and cloud variables. Is the plan to
keep support for the existing scratch cloud variable framework to support
legacy projects? Just wondering :)

Griffpatch


On 2 Nov 2016 00:47, "Aka DadOfMrLog" <notifications@github.com> wrote:

There's no way to detect read-only mode for sure unless another user sends
back a response to your changes (which you can then test for). –And that's
not possible if you're currently the only player, so you *really* cannot
tell in that case.

And if you (think you) are the only player, and you're actually in
write-only mode but don't know it, then when another player joins, you
would never know they have joined, and you still cannot know which mode you
are in (read-only, write-only, or read-write).

So that all means if the project thinks it is currently the only player in
the game, it has no way to know if it is correct about that, or if it is
sending changes which are ruining other players it cannot see (write only),
or if it is totally invisible to other players who may be trying to join on
its slot (read only) – unless, in that last case, it happens to 'catch' a
value on its own cloudvar that it knows it did not set it to (and being
able to catch that sort of thing actually requires that you *don't* set
your own cloudvar too frequently!)
------------------------------

I haven't looked at exactly how your engine works, so can't say for sure,
but it'd be worth you considering how situations such as those above will
play out...

It'd also be worth you working through how your engine would deal with even
the not-so-weird connection cases, where there are potential race
conditions. For example, two users trying to join at very nearly the same
time (on the same 'slot'?), but one or both have slightly bad/slow 'net
connection for some reason (e.g. frequent latency of over half a second
before seeing updates from others). –Is it possible both users could miss
that the other is also trying to join on the same slot (so causing some
chaos)?
------------------------------

Now, you brought up the fallback polling detection... I came across that
capability more recently – after I got fed up trying to make a reliable
(yet easy-to-use) multiplayer engine – and you rightly point out that it
changes things (since I think the write-only case can only happen in
polling mode).
It could be worth revisiting this with that in mind...
------------------------------

Anyway, all of above seems to be somewhat off-topic now, so I'll just say
(at the risk of ruining it!) that I'm really pleased to see how good cloud
has been looking over the last ~30hrs. (Has anything changed?)

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/LLK/scratch-flash/issues/1211#issuecomment-257742115>,
or mute the thread

**Comment by encloinc** [comment]
@griffpatch yes I am aware of how current Multiplayer Projects spam the server with cloud var changes, and I am working on an easy to use engine to fix that. The only way I can think off to possibly detect read or write mode is the following (including some more scenarios and how we should go about fixing them)
**Write Only Mode**
_If your client suddenly thinks 2 or more players stop changing their ping at the same time, your probably in write mode, however this method might be wrong, so you make the client wait another 3 seconds and if everyone is still offline, it will kick itself out (However this can fail on the extreme condition everyone does leave at the same time, but the chances of that are so incredibly small that we shouldn't even worry about that)_
**Read Only Mode**
_To detect this, if one client stops changing their ping and another user catches it, it will send a request to the client in read only mode telling them to stop their game, and since the client can read the var changes, it will stop its game_
**Two or more players join at the same time**
_You could have a cloud var that increases the value by 1 when people are joining the game and decreases the value by 1 when people have entered the game, if the cloud var ever surpasses the value of "1", all the people joining will automatically kick themselves out and set the var to 0 _
**HTTP Fallback**
_The Log Father has made multiple great ways of detecting this, kudos to him _
**Someone is joins the exact time a player leaves**
_This could be bad because lets say id 4 leaves and id 5 comes and takes the spot, well since id 5 hasn't taken the spot of id 4 and the person joining is running down a list of people who are inactive, the person joining will join id 4 leaving id 5 in id 5, and since id 5 detected id 4 not changing, it will too go into id 4, causing chaos. To fix this we could have the person joining check the id that's inactiv AND the id before it checking that there is no one ready to take it, then if it finds someone in the id before the inactive id that is active, it will ignore that inactive id and not take it (Its hard to explain)_
**Not a Scratcher**
_changing a cloud var and checking if the value changed_
**No account**
_Check if there is a username_

_**Im sure there are more scenarios, multiplayer is unpredictable. These are the most common ones I could think of and there fixes**_

**Comment by TheLogFather** [comment]
OK... I've been reluctant to say anything in case I 'jinx' it... but...
Cloud has been behaving way way way better for over a week. :)
So...
Has traffic somehow dropped off? Are there fewer 'bad' cloud projects now? Maybe less (unwitting) abusing of the API in some way? Or maybe you've discovered some way to better regulate traffic...?

**Comment by griffpatch** [comment]
Hello all,

Issue: "Unable to connect to cloud server except through http fallback as
of last night"

Just wanted to make people aware that there might be another problem at the
moment with the cloud server? @TheLogFather's projects are reporting a good
status, yet players of cloud games are not able to connect using the fast
cloud connections, but are 'always' using http fallback mode. I had the
same problem last night and could not connect to any project except through
cloud fallback? Is this a new issue, or something that has occurred before?

Griffpatch

**Comment by TheLogFather** [comment]
Yes, I can confirm that I cannot get through to cloud via port 531. –Probably needs a restart...

None of my cloud clients have been connecting for the past few hours. When I use https://scratch.mit.edu/projects/65451688/ and https://scratch.mit.edu/projects/96798133/, they both show fallback all the time.

**Comment by colbygk** [comment]
Hi,

Thanks for notifying us. In the past week the main primary Scratch web site has been experiencing some problems serving content that we believe are related to the cloudvars service creating congestion in a key part of our infrastructure.  We are currently testing a few methods to alleviate that congestion in preparation for this year's hour of code.

This is what you're seeing with respect to the TCP based cloud variables.

**Comment by griffpatch** [comment]
Since most cloud projects broadcast messages constantly (due to scratchers
not knowing this is a bad idea, or how to do things differently in most
cases), if all cloud services are now limited to cloud fallback mode, then
the majority of real time cloud projects will not operate at all (because
you can't send and receive cloud data if you send data rapidly in http
fallback mode). It might be worth mentioning the fact on the scratch page
so that people understand why their projects will not work over the coming
time-frame? Maybe ;) - Just a thought.

**Comment by colbygk** [comment]
As I mentioned, we are currently exploring various options to alleviate load coming into Scratch infrastructure during that coming time period, if needed.

**Comment by griffpatch** [comment]
I'm not saying you are wrong to do that, on the contrary, that great...
Only it may be worth letting scratchers know (who otherwise will assume
it's their fault)

:)

**Comment by CosmicWebServices** [comment]
I just made a new forum on AT maybe you guys can help contribute with your ideas? https://scratch.mit.edu/discuss/topic/227725/?page=1#post-2310588

I tried to get everything on this issue but I was not 100% sure that I caught all of them

-IcyCoder

**Comment by TheLogFather** [comment]
Cloud varserver URLs are currently unreachable.
(See https://github.com/LLK/scratch-flash/issues/1243#issuecomment-263459992)

Up to now I thought it wasn't particularly important since projects don't use it, and appear to be accessing and changing cloudvars just fine.

However, I've just realised that's only the case for those who can make the connection to cloud through port 531. But fallback polling uses the varserver URL, so anyone who cannot access port 531 is currently unable to use cloud.

**Comment by TheLogFather** [comment]
No cloud at all now – port 531 failing to connect...

**Comment by TheLogFather** [comment]
I'm finding that connection to cloud for projects is only via fallback polling at the moment, and port 531 connection is failing.

As a consequence when using the cloud API, I can't change any cloudvar values (it only works for projects because they are connected via polling).

**Comment by encloinc** [comment]
Well, I think a good idea would be enforcing guidelines about how much cloud vars changes you can make per second

**Comment by Sigton** [comment]
That would be hard to enforce, and would really confuse Scratcher's testing out cloud for the first time. As griffpatch said, they'd be setting many cloud vars many times a second because they don't know of a more efficient and server-friendly method. Suddenly them being not allowed to do that...?

Sigton

**Comment by cwillisf** [comment]
We've made a few changes recently to help deal with the extra load we expect during this year's Hour of Code, which starts in a few days.

Right now the direct TCP connection is turned off, so as @TheLogFather pointed out cloud variables are only working through the HTTP fallback method. We've also increased the polling interval, so the HTTP fallback code is polling for cloud variable values less often than it did before.

We're still thinking about longer-term solutions but we needed to put something in place for the short term. While we explore our options for long-term fixes we'll of course do our best to make the cloud variable system friendly and useful to all Scratchers who try to use it -- whether they're new to Scratch or having conversations about cloud variable internals on GitHub! ;)

**Comment by DavidBuchanan314** [comment]
Maybe you could detect when a project only has 1 concurrent user (which is probably the majority of the time), and only send updates to the API at the end of a session?

**Comment by TheLogFather** [comment]
Since cloud is currently fallback-only, my cloud status project was always showing "Failing".

I've updated the cloud client that controls it to now detect that situation (it then updates cloud values through HTTP), and added a "Fallback Only" status to the display.

Hoping it might be useful to some who might want to keep an eye on how cloud is behaving...

https://scratch.mit.edu/projects/119629398

**Comment by MegaApuTurkUltra** [comment]
This probably breaks most multiplayer games...

**Comment by PolyEdge** [comment]
I know, I know ;-;

**Comment by encloinc** [comment]
I think it's a unwise move to render most cloud dependencies unusable without any sort of warning. Wouldn't it be better to keep the failing cloud servers running rather than moving all cloud based projects to http fallback mode?

**Comment by liam4** [comment]
How unwise that is is debatable considering very, *very* few actual multiplayer projects exist. It's slightly upsetting to the community who does use cloud somewhat extensively for their projects, but at least they're working on the cloud system right now so we can expect improvements in the future. :)

**Comment by colbygk** [comment]
You may notice some improvement to the fallback part of the system...

**Comment by TheLogFather** [comment]
I find it's hard to tell much difference with fallback, unfortunately. Maybe it's a bit less erratic, in terms of consistency between updates?

Still, https://github.com/LLK/scratch-flash/issues/976 means it is practically unusable at the moment for any Scratcher attempting to create or use any kind of multiplayer game. :(

**Comment by TheLogFather** [comment]
Actually, I take back what I said above (the first part)...

It's hard to see a clear difference in terms of receiving cloudvar updates in a project, but now that I look at the logs from the cloud clients I have which are *setting* cloudvars via fallback, I see the response times have become more consistent – not getting erratic delays when changing a cloudvar via fallback.

That's nice. :)

**Comment by PolyEdge** [comment]
I don't see why relaying data across websockets would be so strenuous, websites like slither.io are doing this with way bigger amounts of data 24/7 with multiple player servers and they maintain a low latency.

**Comment by encloinc** [comment]
I have to agree with poly edge on this one. There is no difference between http fallback and a TCP connection in terms of strain. The only problem I see is that people may have overused the TCP connection, but that's no reason to shut it off completely.  As I said earlier, Id prefer a failing TCP connection over a forced http fallback connection.

**Comment by TheLogFather** [comment]
@encloinc: There's a significant difference...

With port 531, the cloud client (i.e. project) only sends something when *it* changes a cloudvar. And it only receives data when someone else changes a cloudvar.

With fallback, the cloud client continually polls the server, and the server always sends back the latest values. –That's why it is restricted to polling only about once per second...

Fallback is generally *way* more verbose then 531 – until someone writes a project which changes the cloudvar(s) dozens of times per second, and each and every change gets sent as a separate update through 531, and then transferred out to everyone else.

----

@PolyEdge: I don't know for sure what the *.io sites are doing (I can't say I've ever played any of them), but I would be surprised if they're sending out more than a handful of updates per second per user. (And that data is probably packed in a much more efficient way, tailored to the way the game needs it.)

Also, I doubt they're recording each and every change, along with who made that change and when, in a database that can then be examined through a web page.

**Comment by CosmicWebServices** [comment]
I know some of them use socket.io also live IM sites use it also.

**Comment by griffpatch** [comment]
Then you will be pleased to know that scratch is already in the process of
transitioning to web sockets (Yay - no more fall back mode), I hope it all
works out super well - I wonder if they will impose broadcast timing limits
over the top. No idea how long until we will see this on the live scratch
page though :). All my cloud games have died a death... Even the old super
stable ones so people have stopped visiting them. :(

On 14 Dec 2016 2:07 p.m., "Cosmic Web Services" <notifications@github.com>
wrote:

I know some of them use socket.io also live IM sites use it also.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/LLK/scratch-flash/issues/1211#issuecomment-267042180>,
or mute the thread

**Comment by encloinc** [comment]
Any updates on this?

**Comment by CatsAreFluffy** [comment]
Bump.

**Comment by PolyEdge** [comment]
Any idea on what's happening with the TCP interface?

**Comment by encloinc** [comment]
Nobody knows. Im sick of fallback to be honest, its too slow

**Comment by MooShoe** [comment]
They said they were porting cloud variables to websockets. I'm not sure 
if they're actually doing that though.

**Comment by ModernFeelGames** [comment]
dude, that'd be sick bro

**Comment by PolyEdge** [comment]
Last time they said anything was in like October and they said the new system would come in January/Febuary this year

**Comment by PolyEdge** [comment]
Also I'm hoping if they even finish this, they'll keep the format the same so that I don't have to rewrite the cloud part of scratchapi yet again 📦

**Comment by liam4** [comment]
@PolyEdge Last time they said anything was about two hours ago, [here](https://github.com/LLK/scratch-flash/issues/1320#issuecomment-307119325):

>We're going to have a new cloud data infrastructure Very Soon™ which will use websockets, so keep an eye on cloud data in the next couple of months!

**Comment by PolyEdge** [comment]
@liam4 Very Soon™ :+1:

**Comment by PolyEdge** [comment]
Also that ninja was insane

**Comment by Hashercsgo** [comment]
3 months later

**Comment by encloinc** [comment]
3 months later

**Comment by jwzimmer** [comment]
The new cloud data work has been deployed as a soft launch (accessible by specific URLs) on Production. 

If anyone would like to help test (which would be greatly appreciated!), here's what you need to know:
- Make sure the URL of the project you are editing ends in `?newcloud`
  - If you load a project, the URL needs to end in (random project number for example) `/projects/171781902/?newcloud`
  - If you are going back and forth between the editor & project page, the URL still needs to have `?newcloud` in it, like  `/?newcloud#editor` or `/?newcloud#player`
  - If you refresh the page or create a new project, the `?newcloud` part needs to be re-added
- Cloud data in the "old"/ existing system on Production is not available in the new soft launch version, but it will be added when the system is rolled out globally (rather than as a soft launch)
  - When you are testing a project with e.g. a high score on Production, that high score won't be available at the `?newcloud` version of the project
  - Data you create during the soft launch of the new cloud data system will be erased when the new work is deployed globally
- The cloud monitor page is accessible at (random project number for example) `/cloudmonitor/171781902/soft/`

Please report any bugs you find here! Thank you.

**Comment by CosmicWebServices** [comment]
Made a small test for it... works great! https://scratch.mit.edu/projects/171867511/?newcloud

**Comment by joker314** [comment]
Could be just me, but this is my console output.

![image](https://user-images.githubusercontent.com/18113170/29588937-0073055c-879c-11e7-8f68-a2a42517d116.png)

It continuously reconnects before giving an error of "WebSocket is already in CLOSING or CLOSED state.", then reconnects again, then gives the same error message, over and over again.

However, **the cloud log correctly updates the variable**. Meaning that despite the errors everything is working well.

So, just error messages -- no actual problem.

I'm using a Windows 10 computer, with my browser being Google Chrome, version 60.0.3112.101

**Comment by griffpatch** [comment]
That's exciting! Do you have any information about this new cloud system...
Is it still throttled to prevent real time gaming or have you been able to
release more bandwidth? Scratch is a wonderful community and I always felt
the we nothing more compelling than releasing a project where you got to
interact with other scratchers...

**Comment by thisandagain** [comment]
@joker314 Thanks for reporting! We'll take a look.

/cc @jwzimmer @colbygk

**Comment by thisandagain** [comment]
@griffpatch The system is still throttled (otherwise we simply would not be able to scale and maintain it) but much more liberally than what is in place right now. The goal is to get everyone moved over to this new system in the short term and then we'll continue to add new features on top of this new infrastructure after Scratch 3.0 is released.

**Comment by joker314** [comment]
(Bug persists when all extensions are disabled)

**Comment by CosmicWebServices** [comment]
On Firefox (latest stable) I get 
Websocket closed, code:1000 reason:   project_base.js:1:5740
07:31:14 PM | wrn | Connection closed to cloud server Object { reconnect: 1649.0207351744175, attempts: 3 }  171867511
07:31:16 PM | inf | Attempt reconnect to cloud server.  171867511
07:31:20 PM | inf | Successfully connected to cloud data server

Every 5 seconds or so

**Comment by ateesdalejr** [comment]
@thisandagain You say that cloud is still throttled very well below the rates required for realtime multiplayer but I have a project: https://scratch.mit.edu/projects/171930267/?newcloud that works quite well with some very basic multiplayer functionality. Has this changed since you posted last?

**Comment by jwzimmer** [comment]
@CosmicWebServices Thanks for reporting that. Did you notice any problems in the functionality of your project when you saw the error? If you have a link to the project that occurred in as well as steps that cause it to happen, that would be helpful, too. 👍

**Comment by CosmicWebServices** [comment]
@jwzimmer no not really the link is already above (the one I posted)


Suggestion: https://scratch.mit.edu/discuss/topic/274278/?page=2#post-2799178

**Comment by colbygk** [comment]
Clouddata has now been migrated to the new websockets based platform and `?newcloud` is no longer required to have a project use it.

**Comment by griffpatch** [comment]
I am seeing quite a few problems with the cloud data over the last week
since I started looking at it again.

I have been seeing issues with a high score table script where it appears
to lose previous players scores.
What appears to happen is that the scratch project must fail to load the
current value from the cloud server, but succeeds in writing the value back
(such that only the scores in the cloud variable at the time I last saved
the project are persisted long term). Also, complete failure to connect
appear to be frequent enough.

Right now I am unable to even get a cloud variable to update in the cloud
log. I update it in the project, but it is not seen by other connected
players and no record is logged.
Chrome's console is logging this:

11:18:09 AM | inf | Connecting to cloud data server {server: "
> clouddata.scratch.mit.edu"}server: "clouddata.scratch.mit.edu"__proto__:
> constructor: ƒ Object()hasOwnProperty: ƒ hasOwnProperty()isPrototypeOf: ƒ
> isPrototypeOf()propertyIsEnumerable: ƒ
> propertyIsEnumerable()toLocaleString: ƒ toLocaleString()toString: ƒ
> toString()valueOf: ƒ valueOf()__defineGetter__: ƒ
> __defineGetter__()__defineSetter__: ƒ __defineSetter__()__lookupGetter__: ƒ
> __lookupGetter__()__lookupSetter__: ƒ __lookupSetter__()get __proto__: ƒ
> __proto__()set __proto__: ƒ __proto__()
> VM949:1 11:18:10 AM | inf | Successfully connected to cloud data server
> VM950:1 11:18:36 AM | inf | Play recorded
> VM951:1 11:19:11 AM | inf | Attempt reconnect to cloud server.
> VM953:1 11:19:14 AM | inf | Successfully connected to cloud data server
> 11:18:09 AM | inf | Connecting to cloud data server {server: "
> clouddata.scratch.mit.edu"}server: "clouddata.scratch.mit.edu"__proto__:
> constructor: ƒ Object()hasOwnProperty: ƒ hasOwnProperty()isPrototypeOf: ƒ
> isPrototypeOf()propertyIsEnumerable: ƒ
> propertyIsEnumerable()toLocaleString: ƒ toLocaleString()toString: ƒ
> toString()valueOf: ƒ valueOf()__defineGetter__: ƒ
> __defineGetter__()__defineSetter__: ƒ __defineSetter__()__lookupGetter__: ƒ
> __lookupGetter__()__lookupSetter__: ƒ __lookupSetter__()get __proto__: ƒ
> __proto__()set __proto__: ƒ __proto__()
> VM949:1 11:18:10 AM | inf | Successfully connected to cloud data server
> VM950:1 11:18:36 AM | inf | Play recorded
> VM951:1 11:19:11 AM | inf | Attempt reconnect to cloud server.
> VM953:1 11:19:14 AM | inf | Successfully connected to cloud data server
> VM954:1 11:20:16 AM | inf | Attempt reconnect to cloud server.
> VM956:1 11:20:17 AM | inf | Successfully connected to cloud data server
> VM957:1 11:21:19 AM | inf | Attempt reconnect to cloud server.
> VM959:1 11:21:20 AM | inf | Successfully connected to cloud data server
> VM960:1 11:22:28 AM | inf | Attempt reconnect to cloud server.
> VM962:1 11:22:33 AM | inf | Successfully connected to cloud data server
> VM963:1 11:23:45 AM | inf | Attempt reconnect to cloud server.
> VM965:1 11:23:46 AM | inf | Successfully connected to cloud data server
> VM966:1 11:24:47 AM | inf | Attempt reconnect to cloud server.


over and over! but it doesn't appear tat it has actually succeeded...

I have been aware of quite a number of times that connectivity appeared to
be down, and I haven't actually been on that many times... so it feels like
it might be quite an issue?

Thanks,

Griffpatch

**Comment by griffpatch** [comment]
Please ignore me... Right now it IS working... I made a schoolboy error ;)
- However, the original issue stands with the score table and periodic
cloud down?

**Comment by thisandagain** [comment]
Thanks @griffpatch. I'm going to move this over to another issue.

**Comment by colbygk** [comment]
@griffpatch Thanks for working on cloud data projects! 

Last night (21 Sep 2017), I deployed some changes to the new cloud data service that has improved connectivity issues.

Could you point me at a project where you're seeing the high score and data integrity issues?

**Comment by griffpatch** [comment]
It was on this project :)

https://scratch.mit.edu/projects/174671513/
